### ISSUE TITLE
Incorrect highlighting

### ISSUE THREAD

**Issue by nad** [agda/agda#4260]
I have noticed that the current development version of Agda sometimes generates incorrect highlighting. My guess is that some range is incorrect, because highlighting ends up where it shouldn't be (for instance in comments). Unfortunately the problem seems to be hard to reproduce. The problem can perhaps be triggered in the following way (using Agda 2.6.1-cf76109):
```
git clone http://www.cse.chalmers.se/~nad/repos/equality
cd equality
git checkout da311d97529ce6e162230fff54833baeecae834f
sed --in-place -e '24d' -e '397,411d' src/List.agda
agda --html src/Fin.agda
```
In the file `html/Fin.html` the comment before `largest` is incorrectly highlighted, and there is also some incorrectly highlighted code in the body of `largest≡Fin↔<`.

If I remove the `.agdai` file for `src/Fin.agda` (but nothing else) and run the final command above again, then the problems mentioned above do not appear in the HTML file. Could it be the case that, in some cases, there is some incorrect data in the state after a file has been type-checked, but that this data is not recreated when information is read from the interface file?

I think this is a regression. However, the code used above is not accepted by Agda 2.6.0.1, so I am not entirely certain.

**Comment by gallais** [comment]
Is this #4123?

**Comment by nad** [comment]
Right, I had forgotten about that ticket. Note that the problem does not only affect the HTML backend.

Do you have a test case suitable for use with `agda-bisect`? My test case is not accepted by Agda 2.6.0.1.

**Comment by nad** [comment]
> Could it be the case that, in some cases, there is some incorrect data in the state after a file has been type-checked, but that this data is not recreated when information is read from the interface file?

The serialiser throws away quite a few ranges, so this seems plausible. Thus a fix of #4316 might lead to a fix of (the symptom of) this problem.

Perhaps we could track down this problem by checking for incorrect ranges in (say) the serialiser.

**Comment by nad** [comment]
I tweaked `mergeAspects` so that it (sometimes) complains when `Bound` is combined with other `NameKind`s, and managed to construct the following example:
```agda
record R : Set where
  constructor c

F : R → Set₁
F = λ (c) → Set
```
Here Agda tries to combine the following two values of type `Aspects`:
```
Aspects {aspect = Just (Name (Just Bound) False), otherAspects = fromList [], note = Nothing, definitionSite = Just (DefinitionSite {defSiteModule = TopLevelModuleName {moduleNameRange = 1,1-4, moduleNameParts = ["Bug"]}, defSitePos = 59, defSiteHere = True, defSiteAnchor = Nothing}), tokenBased = NotOnlyTokenBased}
```
```
Aspects {aspect = Just (Name (Just (Constructor Inductive)) False), otherAspects = fromList [], note = Nothing, definitionSite = Just (DefinitionSite {defSiteModule = TopLevelModuleName {moduleNameRange = 1,1-4, moduleNameParts = ["Bug"]}, defSitePos = 36, defSiteHere = False, defSiteAnchor = Just "c"}), tokenBased = NotOnlyTokenBased}
```
I'm guessing that at least one problem has something to do with the new record pattern lambdas.

**Comment by nad** [comment]
I have managed to construct a small test case. However, it behaves a bit differently from the OP, so perhaps it is caused by a different problem.

The test case uses two modules:
* `A.agda`:
  ```agda
  module A where

  postulate
    F : Set → Set

  syntax F X = G X
  ```
* `B.agda`:
  ```agda
  module B where

  open import A

  --          syntax F X = G X

  H : Set → Set
  H X = G X
  ```
The string `syntax F X = G X` is located at the same position in the two files. If `B.agda` is loaded once, then the `G` in the comment is highlighted as a postulate, and if the file is reloaded, then the two `X`s are also highlighted as bound variables. (Actually, the `X`s first get the `face` text property `(agda2-highlight-bound-variable-face font-lock-comment-face)` and then `(agda2-highlight-bound-variable-face)`, whereas the `G` first gets the `face` text property `(agda2-highlight-postulate-face font-lock-comment-face)` and then `(agda2-highlight-postulate-face)`.)

The problem does not seem to be related to caching or interactive highlighting.

Here is a test program for use with `agda-bisect`:
```haskell
{-# LANGUAGE RecordWildCards #-}

import Data.List
import System.Exit

import RunAgda

main :: IO ()
main = runAgda [ "--no-libraries"
               , "--ignore-interfaces"
               ] $ \(AgdaCommands { .. }) -> do

    -- Discard the first prompt.
    echoUntilPrompt

    -- Load the module.
    let f = "B.agda"
    send $ command "load" f (Just "NonInteractive Direct")
                   (Just $ show f ++ "[]")
    output <- echoUntilPrompt

    -- Check the output.
    if "*Error*" `isInfixOf` unlines output
     then exitWith (ExitFailure 127)
     else if "'(53 54 (bound)) '(57 58 (postulate)) '(59 60 (bound))"
               `isInfixOf`
             unlines output
     then exitFailure
     else exitSuccess
```

**Comment by nad** [comment]
Bisection points to 3d1885716cf22779db038dccb3944d6b82a57760 ("[ fix #758 ] highlighting for syntax declarations"). However, I can imagine that the underlying problem was introduced in the previous commit (42e8b56e9beb84f6d26a08779b7d49934953fc6c, "[ #758 ] refactor: add ranges to notations").

**Comment by nad** [comment]
Consider the following code:
```agda
open import Agda.Builtin.Sigma

F : Σ Set (λ _ → Set) → Set
F = λ (A , B) → B
```
If I invoke "go-to-definition" on the occurrences of `,` and `B` in the pattern I end up at the `A`. However, for the `B` in the right-hand side I end up at the `B` in the pattern.

**Comment by nad** [comment]
The following code seems to be handled correctly:
```agda
F : Σ Set (λ _ → Set) → Set
F P = let (A , B) = P in B
```
This suggests that the problem discussed [above](https://github.com/agda/agda/issues/4260#issuecomment-571567480) is related to the new record pattern lambdas. @gallais, can you take a look at this?

**Comment by nad** [comment]
I've moved the record pattern lambda problem to a new issue (#4356). @UlfNorell, can you take a look at the issue related to syntax declarations? Hopefully that problem is the cause of the issues in the OP and #4123.

**Comment by nad** [comment]
@UlfNorell, in 42e8b56e9beb84f6d26a08779b7d49934953fc6c you added ranges to `GenPart`. Your implementation of `setRange` does not change the range of `IdPart`s:

https://github.com/agda/agda/blob/4e6ef5bfc9a2a516c103384fbc348d8eb8780137/src/full/Agda/Syntax/Notation.hs#L87-L92

Was this intentional? There is no comment that states that it was.

(I'm not claiming that this is the cause of the bug.)

**Comment by UlfNorell** [comment]
That looks like a mistake.

**Comment by nad** [comment]
This code looks suspicious to me:

https://github.com/agda/agda/blob/4e6ef5bfc9a2a516c103384fbc348d8eb8780137/src/full/Agda/Interaction/Highlighting/Generate.hs#L942-L964

It appears to me as if the notations in names contain the ranges of the notations' *definitions*, not their uses. I patched the problem by using the technique by @andreasabel that can be seen in the code above, and a quick test suggests that this works. However, I suggest that we abandon the current hack, which involves generating highlighting for a fixity/syntax declaration *every time the name is used*, and that we instead generate this highlighting only once per declaration.

@andreasabel asks why the ranges are serialised. I don't think they are. As noted above the problem disappears when state is read from interface files.

**Comment by nad** [comment]
> That looks like a mistake.

I've opened a new issue for this thing (#4362).

**Comment by nad** [comment]
My quick fix seems to fix the problem in the OP and the problem in #4123.

**Comment by nad** [comment]
@UlfNorell, you replaced `__IMPOSSIBLE__` with `mempty` in the following code more than nine years ago (8992c8b2d542743a84ff2db21c8857aaf1c5d538):

https://github.com/agda/agda/blob/4e6ef5bfc9a2a516c103384fbc348d8eb8780137/src/full/Agda/Interaction/Highlighting/Generate.hs#L863-L870

Do you remember why? I tried to replace `mempty` with `__IMPOSSIBLE__`, and encountered no problems when I type-checked one of my libraries.

**Comment by UlfNorell** [comment]
I'm pretty sure I did it because it was triggered. It might not be anymore though, so changing back to `__IMPOSSIBLE__` could be worthwhile.

**Comment by nad** [comment]
> I'm pretty sure I did it because it was triggered. It might not be anymore though, so changing back to `__IMPOSSIBLE__` could be worthwhile.

I suggest that we wait until after the upcoming release.

**Comment by andreasabel** [comment]
Is it still a `regression on master`?  If not, please clear this tag. We shouldn't release fresh known regressions, if we can help.

**Comment by nad** [comment]
If you want to close the issue, please open a new issue related to the following comment:

> However, I suggest that we abandon the current hack, which involves generating highlighting for a fixity/syntax declaration *every time the name is used*, and that we instead generate this highlighting only once per declaration.

**Comment by andreasabel** [comment]
I bumped the milestone since the `regression on master` has been solved.

**Comment by andreasabel** [comment]
> However, I suggest that we abandon the current hack, which involves generating highlighting for a fixity/syntax declaration _every time the name is used_, and that we instead generate this highlighting only once per declaration.

The problem is that fixity declarations are only represented in concrete syntax (where the name kind has not been computed yet) and not in the abstract syntax (where we know the name kind).  A systematic fix would add such declarations to the abstract syntax (and necessarily, also to the nice syntax in Concrete.Definitions), for the purpose of highlighting.  There already are certain declarations in the abstract syntax (like `Open`) only for highlighting purposes.

**Comment by UlfNorell** [comment]
I don't see any harm in the current "hack", and it certainly doesn't seem worth the effort to change it. If someone feels strongly about this please open a new issue with a cost/benefit analysis.

**Comment by andreasabel** [comment]
> I bumped the milestone since the `regression on master` has been solved.

Undoing that since the issue was closed without any further action after 2.6.1.
